### DISCUSSION TITLE
Flet VS Pynecone

### DISCUSSION THREAD

**Discussion by saurabhwadekar** [flet-dev/flet#1011] [General]
Flet vs Pynecone who is best

**Comment by FeodorFitsner** [comment]
Really? :)

There is no "the best" framework - that's why there are thousands of UI frameworks in every ecosystem. Each framework has its own purpose.

* Flet is for making desktop, web and (soon) mobile apps while Pynecone is for web only.
* Flet UI is based on Flutter, which could feel a bit "heavier" and "painted" but the great benefit is that it looks the same on every platform. Pynecone uses React components. While there are some similarities with web Flutter is trying to take a different approach to building UIs: no CSS, a single "virtual pixel" measurement unit instead of px, pt, rem, %, etc. in HTML, more strict and precise top-to-bottom layout engine. Flutter has awesome animation too.
* Pynecone apps have "declarative" programming model similar to React with centralized state management, while Flet implements "imperative" model (with the state stored in app controls) which beginners may find more intuitive.
* Flet apps can be [published as a static web site](https://flet.dev/docs/guides/python/publishing-static-website), hosted on a free hosting and run entirely in the browser. Pynecone app requires a hosting being able to run Python code.
* Flet app can be [packaged into standalone desktop app bundle](https://flet.dev/docs/guides/python/packaging-desktop-app) - a perfect memory-effective alternative to Electron!

In general, if you need to build a website-like web app and you are familiar with web technologies (CSS, React state management) - use Pynecone. If you want to build a multi-platform app that your users will be running on desktop, web and mobile use Flet.

**Comment by hololeo** [comment]
i have taken both for a test drive, more flet than pynecone and i have a few observations

1. what flet does to dart/flutter, pynecone does for react/next.js. You absolutely do not need to know any react to produce react from python code. much like with flet you only need to know python to 'generate' dart/flutter apps
2. pyncone is bound to its server to run much like flet was in the beginning. Eventually this will be made serverless as well.
3. i find pyncone ideal for stiching together a web app with different technologies (including flet). whereas flet is one big container on the page (the dart flutter.js canvas), pynecone is made up of individual html react components. So pynecone is ideal for wrapping up flet in an otherwise html app (flet is a bit like applets and flash embeds)
4. a lot of great ideas from pynecone's api design should make it into flet. its easier to style components in PC for example..but that is because they get the entire css at their command whereas with flet you use whatever flutter layout/styling is under it.
5. overall they are both excellent to learn and enjoy. Notice i said 'enjoy'. It doesnt feel like work when using Flet or Pynecone ... its fun! 
6. with that said, if i was exilled to on an island and i could only take one with me, id take Flet with me. Flet Forevaa!
